### DISCUSSION TITLE
0067 XLS-67d: Charge

### DISCUSSION THREAD

**Discussion by tequdev** [XRPLF/XRPL-Standards#194] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by xVet** [comment]
Great proposal Tequ! Question, what's the difference to a Payment transaction? I thought you would do a pull payment and take `Amount` from the `destination` but you describe it as just a payment to `destination`

**Reply by dangell7** [reply]
It would go on the transaction. Example:

```
{
  "TransactionType": "NFTokenMint",
  "NFTokenTaxon": 0,
  "URI": "DEADBEEF",
  "Flags": 1,
  "Charge":  {
      "Amount": "1000000",
      "Destination": "rH4KEcG9dEwGwpn6AyoWK9cZPLL4RLSmWW"
  }
}
```

Where the Charge destination is the UI platform. Like a service fee.

**Reply by xVet** [reply]
I see, thanks for the example, made it more visible!

**Comment by shefi** [comment]
Great proposal, Tequ!

It should help those building the UI tremendously. If they all can make money, they will have enough capital to push the marketing far and wide for their offerings on XRPL. This way new engagement will be brought in to XRPL and help with the overall growth of the ecosystem.

I would like to suggest for the Charge field to be an array to allow for multiple amount and destination pairs. Array or any appropriate type for rippled codebase, because I don't have knowledge of it. I think I saw STArray somewhere.

Why?

It could be useful if there will be a time that a non-profit bounty reward system will come to life as described in this [X Post](https://twitter.com/HatLord/status/1781424785056010311) I have shared. UIs could give their users an option for them to donate an amount (however small) to any such non-profit they want to donate to.

It could be something like Kyber Swap has with their tipping feature for DEX Screener on this screenshot for example:
![Screenshot_20240422_194107_Chrome.jpg](https://github.com/XRPLF/XRPL-Standards/assets/477494/ab2a35e4-092c-4c72-9d70-00012caec4f4)

Here is the [link](https://kyberswap.com/partner-swap?chainId=10&inputCurrency=0x94b008aA00579c1307B0EF2c499aD98a8ce58e58&outputCurrency=0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85&clientId=dexscreener&feeReceiver=0x0DA2a82ED2c387d1751ccbAf999A80b65bdb269E&enableTip=true&chargeFeeBy=currency_out&feeAmount=10) to that interface.

It should all be optional, but if they build right their users will appreciate the value they are getting and will definitely choose to pay either the fees, donations, or both.

I hope this proposal gets accepted and goes live. It will help a lot all those developing for XRPL to monetize their products so they can survive and keep pushing more and better. As well as maybe help with funneling those community donations for further developments of the core ledger or other ecosystem solutions.

**Comment by EnchStyle** [comment]
Can you provide a straightforward explanation of what this feature does and how it improves upon XLS-56d, without using technical jargon? The abstract on itself does not answer my question or help me to understand the perspective. :)
Thanks

**Reply by shefi** [reply]
Look at the screenshot I have attached above.

If I understand it correctly, the current proposal if implemented would allow you to add something like the part I have pointed the arrow to in the screenshot. So you can collect fees from your users on each transaction.

My suggestion is to allow for multiple such options, so you can show one for your fees and maybe one for donations at the same time.

I hope that helps.

**Reply by sappenin** [reply]
@shefi couldn't your screenshot be enabled using [Batch transactions](https://github.com/XRPLF/XRPL-Standards/discussions/162)? E.g., if a user chooses to add a tip, then this just becomes another transaction in a batch.

**Reply by shefi** [reply]
It might very well be possible. I don't have much knowledge of the code base. I am just a lurker, unfortunately.

Maybe the pros and cons of doing it with batch transactions vs doing it with how Tequ is proposing should be discussed.

Furthermore, if something like this can be done with batch transactions, I suppose what Tequ is proposing could also be done that way?

**Reply by mvadari** [reply]
I'd personally rather not have multiple solutions for a single problem.

**Reply by shefi** [reply]
Yes, if this solution is solvable with batch transactions, that should be used.

Do you see any pros and/or cons doing this way vs doing it the batch transaction way?

I apologize beforehand if I am adding more unnecessary workload for you all, because of my lack of knowledge of the codebase.

**Reply by mvadari** [reply]
You can take a look at the spec here: https://github.com/XRPLF/XRPL-Standards/discussions/162

**Reply by shefi** [reply]
I will try to take a look at it and see if I can follow tomorrow, because I am on a mobile device now and it's not easy to follow the flow.

I am shooting blind here, but one drawback I may foresee is the need for paying the transaction fee for each transaction in the batch vs paying only once this way? I could be wrong though, and please correct me if I am. This is how I learn as well.

**Reply by mvadari** [reply]
Yes, that's the primary difference I see as well - but I'm not sure that that's worth a whole different spec to solve the same problem.

**Reply by shefi** [reply]
Yes, it's not necessarily bad, since it helps with the deflationary economics of the native token as well.

However, if there is any kind of workaround, while maintaining the same quality requirements, that can be done to avoid paying extra fees, it might be worth to explore that route.

**Reply by EnchStyle** [reply]
@mvadari maybe you could kindly provide more clarification? as you clearly understand XLS-56d better than anyone and can compare to this. As I am still not sure if I understand the nuances and differences between two.

**Comment by rothexD** [comment]
should we add a `[optional]DestinationTag` to the `charge` field?

**Reply by shefi** [reply]
Yeah, that's also great suggestion. Destination tag should be in for sure.

**Comment by Panosmek** [comment]
Great proposal that needs to pass asap! Is it possible to charge a percentage instead of an amount? If a UI wants to charge 0.1% on each transaction for example, can it be done?

**Reply by mvadari** [reply]
You can just calculate that when you generate the transaction.

**Comment by mvadari** [comment]
Is there a method to support multiple charges on one transaction, such as if the UI wants to charge a fee and the wallet also wants to charge a separate fee?

**Reply by tequdev** [reply]
As others have pointed out, changing the field to an array would allow for multiple charges.

**Comment by tequdev** [comment]
The Batch functionality, which has some overlap with this proposal, is currently under development, and I don't see this proposal moving forward to the next phase at this time.

The unique point of this feature, as opposed to the Batch, is that it is clear from the presence of the Charge field that the amount will be collected as a fee.

If anyone would like to move this feature forward, please let me know. I will be happy to help.

**Comment by mvadari** [comment]
@tequdev should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.
